### PR TITLE
SOLR-16693: timeAllowed: use more efficient TimeLimitingBulkScorer

### PR DESCRIPTION
**Pull request by dsmiley**
Instead of ExitableDirectoryReader

https://issues.apache.org/jira/browse/SOLR-16693

TODO... maybe allow use of ExitableDirectoryReader via sys prop?  BTW there are code paths using the reader that wouldn't use the search method yet may have a timeAllowed param.  I think I've seen this in an erroneous (bug or accident; haven't root caused yet) scenario at work but could be valid.

Wish we had benchmarks on this.  Any way, it's interesting that all tests pass so easily :-)

### CODE REVIEW

**Review by mkhludnev — APPROVED** [review]

File: solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java [hunk 1]
```diff
@@ -713,6 +713,33 @@ public QueryResult search(QueryResult qr, QueryCommand cmd) throws IOException {
     return qr;
   }
 
+  @Override
+  protected void search(List<LeafReaderContext> leaves, Weight weight, Collector collector)
+      throws IOException {
+    final var queryTimeout = SolrQueryTimeoutImpl.getInstance();
+    if (queryTimeout.isTimeoutEnabled() == false) {
+      // no timeout.  Pass through to subclass
```
Comment (cpoerschke) [inline solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:721]:
```suggestion
      // no timeout.  Pass through to super class
```

File: solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java [hunk 2]
```diff
@@ -713,6 +713,33 @@ public QueryResult search(QueryResult qr, QueryCommand cmd) throws IOException {
...
+      // So we need to make a new IndexSearcher instead of using "this".
+      new IndexSearcher(reader) { // cheap, actually!
+        void searchWithTimeout() throws IOException {
+          setTimeout(queryTimeout.makeLocalImpl());
+          super.search(leaves, weight, collector); // FYI protected access
+          if (timedOut()) {
+            // We might not even be using ExitableDirectoryReader yet various places in Solr check
+            //  for this exception.  Sometimes others but this seems most suitable.
+            throw new ExitableDirectoryReader.ExitingReaderException(
+                "Timed out.  Not actually via ExitableDirectoryReader");
```
Comment (cpoerschke) [inline solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:736]: Maybe an alternative might be to have a `class SolrNotActuallyExitableDirectoryReader extends ExitableDirectoryReader.ExitingReaderException` approach which would be thrown here and then over time the various places could be switched over to check for `SolrNotActuallyExitableDirectoryReader` instead or to have the checks phased out.

**Review by cpoerschke — APPROVED** [review]

File: solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java (hunk 2 above)
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:736]: I assume you're not serious about the name SolrNotActuallyExitableDirectoryReader.  Two name proposals:
* TimeAllowedExceededException -- generic.  But despite it being generic, the reality is that there are other exceptions that are effectively similar that would still exist, so maybe give up on trying to be generic?
* TimeAllowedExceededFromScorerException -- more specific as to what makes this timeout what it is.

Comment (cpoerschke) [reply solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:736]: > I assume you're not serious about the name SolrNotActuallyExitableDirectoryReader. Two name proposals: ...

Yeah, that was not meant to be a serious name suggestion but only a way to convey the `class ... extends ExitableDirectoryReader.ExitingReaderException` idea.

`TimeAllowedExceededException` vs. `TimeAllowedExceededFromScorerException` -- the latter to me seems closer to the current `ExitableDirectoryReader.ExitingReaderException` which is also pretty specific.

Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java]: Good; I chose the latter; PR updated.
I think it'd be worthwhile for Lucene to have a base class of these exceptions that is public, therefore Lucene apps can catch one general exception no matter which approach is being taken.

**Comment by dsmiley** [comment]
I plan to merge end of Monday or so.

**Pull request merged**